File: Ex2/Ex2.js
const morseAlphabet = [
    ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---",
    ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."
];

// Map to store the correspondence between characters and their Morse representation.
const charToMorse = new Map();

function initCharToMorseMap() {
    for (let i = 0; i < 26; ++i) {
        charToMorse.set(String.fromCharCode(97 + i), morseAlphabet[i]);
    }
}

// Function to generate permutations of the given string and add them to the container.
function generatePermutations(str, l, r, permutations) {
    if (l === r) {
        permutations.add(str);
    } else {
        for (let i = l; i <= r; i++) {
            str = swap(str, l, i);
            generatePermutations(str, l + 1, r, permutations);
            str = swap(str, l, i);
        }
    }
}

function swap(str, i, j) {
    let charArray = str.split('');
    let temp = charArray[i];
    charArray[i] = charArray[j];
    charArray[j] = temp;
    return charArray.join('');
}

// Function to convert a string to Morse code using the created map.
function stringToMorse(str) {
    let morseString = "";
    for (let i = 0; i < str.length; i++) {
        morseString += charToMorse.get(str[i]);
    }
    return morseString;
}

function main() {
    initCharToMorseMap();
    let input = prompt("Enter words to permute (separate them with spaces): ");

    // Split the input into separate words.
    let words = input.split(' ');
    let uniqueMorseWords = new Set();

    for (let word of words) {
        let permutations = new Set();
        generatePermutations(word, 0, word.length - 1, permutations);

        for (let permutation of permutations) {
            uniqueMorseWords.add(stringToMorse(permutation));
        }
    }

    console.log("Unique words in Morse code:");
    for (let morseWord of uniqueMorseWords) {
        console.log(morseWord);
    }

    console.log("Number of unique words: " + uniqueMorseWords.size);
}

main();

File: Ex2/Ex2.py
import itertools

morse_alphabet = [
    ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---",
    ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."
]

# dict для хранения соответствия между символами и их представлением в коде Морзе.
char_to_morse = {}

def init_char_to_morse_map():
    for i, code in enumerate(morse_alphabet):
        char_to_morse[chr(ord('a') + i)] = code

# Функция генерирует перестановки переданной строки и добавляет их в контейнер.
def generate_permutations(str):
    return set(''.join(p) for p in itertools.permutations(str))

# Преобразую строку в код Морзе, используя созданный dict.
def string_to_morse(str):
    return ''.join(char_to_morse[c] for c in str)

def main():
    init_char_to_morse_map()
    input_str = input("Введите слова для перестановки (разделяйте их пробелами): ")

    # Разделяем ввод на отдельные слова.
    words = input_str.split()
    unique_morse_words = set()

    for word in words:
        permutations = generate_permutations(word)

        for permutation in permutations:
            unique_morse_words.add(string_to_morse(permutation))

    print("Уникальные слова в языке Морзе:")
    for morse_word in unique_morse_words:
        print(morse_word)

    print("Количество уникальных слов:", len(unique_morse_words))

if __name__ == "__main__":
    main()